### PR TITLE
Limit backfill validation concurrency

### PR DESCRIPTION
**Pull request by aldernero**
#### What this PR does

This PR adds a configurable limit to the compactor that sets the maximum number of concurrent backfill validations. This can be used to prevent compactor resources from becoming over-utilized due to the validation phase of block upload which consumes both local storage and CPU resources.

- adds a new compactor configuration parameter `-compactor.max-block-upload-validation-concurrency`, defaults to 1, 0 means unlimited
- adds a new gauge `cortex_block_upload_current_validations` to monitor concurrency

#### Checklist

- [X] Tests updated
- [X] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by aldernero** [comment]
I tested that mimirtool functions as expected with the `development/mimir-microservices-mode/docker-compose.jsonnet` and some tests blocks:

### first client starts, completes normally
```
❯ mimirtool backfill --log.level="debug" 01GKN2SS8BP9G2P7K2NWBKVC5W
INFO[0000] log level set to debug                       
INFO[0000] Backfilling                                   blocks=01GKN2SS8BP9G2P7K2NWBKVC5W user=anonymous
DEBU[0000] New Mimir client created                      address="http://localhost:8006" id=anonymous
INFO[0000] making request to start block upload          block=01GKN2SS8BP9G2P7K2NWBKVC5W file=meta.json path=01GKN2SS8BP9G2P7K2NWBKVC5W
DEBU[0000] sending request to Grafana Mimir API          method=POST url="http://localhost:8006/api/v1/upload/block/01GKN2SS8BP9G2P7K2NWBKVC5W/start"
DEBU[0000] checking response                             status="200 OK"
INFO[0000] uploading block file                          block=01GKN2SS8BP9G2P7K2NWBKVC5W file=index path=01GKN2SS8BP9G2P7K2NWBKVC5W size=34211534
DEBU[0000] sending request to Grafana Mimir API          method=POST url="http://localhost:8006/api/v1/upload/block/01GKN2SS8BP9G2P7K2NWBKVC5W/files?path=index"
DEBU[0000] checking response                             status="200 OK"
INFO[0000] uploading block file                          block=01GKN2SS8BP9G2P7K2NWBKVC5W file=chunks/000001 path=01GKN2SS8BP9G2P7K2NWBKVC5W size=207314021
DEBU[0000] sending request to Grafana Mimir API          method=POST url="http://localhost:8006/api/v1/upload/block/01GKN2SS8BP9G2P7K2NWBKVC5W/files?path=chunks%2F000001"
DEBU[0001] checking response                             status="200 OK"
DEBU[0001] sending request to Grafana Mimir API          method=POST url="http://localhost:8006/api/v1/upload/block/01GKN2SS8BP9G2P7K2NWBKVC5W/finish"
DEBU[0001] checking response                             status="200 OK"
DEBU[0001] sending request to Grafana Mimir API          method=GET url="http://localhost:8006/api/v1/upload/block/01GKN2SS8BP9G2P7K2NWBKVC5W/check"
DEBU[0001] checking response                             status="200 OK"
DEBU[0001] checked block upload state                    block=01GKN2SS8BP9G2P7K2NWBKVC5W path=01GKN2SS8BP9G2P7K2NWBKVC5W state=validating
DEBU[0021] sending request to Grafana Mimir API          method=GET url="http://localhost:8006/api/v1/upload/block/01GKN2SS8BP9G2P7K2NWBKVC5W/check"
DEBU[0021] checking response                             status="200 OK"
DEBU[0021] checked block upload state                    block=01GKN2SS8BP9G2P7K2NWBKVC5W path=01GKN2SS8BP9G2P7K2NWBKVC5W state=complete
INFO[0021] block uploaded successfully                   block=01GKN2SS8BP9G2P7K2NWBKVC5W path=01GKN2SS8BP9G2P7K2NWBKVC5W
INFO[0021] finished uploading blocks                     already_exists=0 failed=0 succeeded=1
```

### second client starts immediately after first, waits and retries
```
❯ mimirtool backfill --log.level="debug" 01GKN2TQV5A3D13852P6CKCS90
INFO[0000] log level set to debug                       
INFO[0000] Backfilling                                   blocks=01GKN2TQV5A3D13852P6CKCS90 user=anonymous
DEBU[0000] New Mimir client created                      address="http://localhost:8006" id=anonymous
INFO[0000] making request to start block upload          block=01GKN2TQV5A3D13852P6CKCS90 file=meta.json path=01GKN2TQV5A3D13852P6CKCS90
DEBU[0000] sending request to Grafana Mimir API          method=POST url="http://localhost:8006/api/v1/upload/block/01GKN2TQV5A3D13852P6CKCS90/start"
DEBU[0000] checking response                             status="200 OK"
INFO[0000] uploading block file                          block=01GKN2TQV5A3D13852P6CKCS90 file=index path=01GKN2TQV5A3D13852P6CKCS90 size=49001032
DEBU[0000] sending request to Grafana Mimir API          method=POST url="http://localhost:8006/api/v1/upload/block/01GKN2TQV5A3D13852P6CKCS90/files?path=index"
DEBU[0000] checking response                             status="200 OK"
INFO[0000] uploading block file                          block=01GKN2TQV5A3D13852P6CKCS90 file=chunks/000001 path=01GKN2TQV5A3D13852P6CKCS90 size=413775399
DEBU[0000] sending request to Grafana Mimir API          method=POST url="http://localhost:8006/api/v1/upload/block/01GKN2TQV5A3D13852P6CKCS90/files?path=chunks%2F000001"
DEBU[0002] checking response                             status="200 OK"
DEBU[0002] sending request to Grafana Mimir API          method=POST url="http://localhost:8006/api/v1/upload/block/01GKN2TQV5A3D13852P6CKCS90/finish"
DEBU[0002] checking response                             status="429 Too Many Requests"
ERRO[0002] response                                      body="too many block upload validations in progress\n" status="429 Too Many Requests"
WARN[0002] will sleep and try again                      block=01GKN2TQV5A3D13852P6CKCS90 error="POST request to http://localhost:8006/api/v1/upload/block/01GKN2TQV5A3D13852P6CKCS90/finish failed: server returned HTTP status: 429 Too Many Requests, body: \"too many block upload validations in progress\\n\"" path=01GKN2TQV5A3D13852P6CKCS90
DEBU[0022] sending request to Grafana Mimir API          method=POST url="http://localhost:8006/api/v1/upload/block/01GKN2TQV5A3D13852P6CKCS90/finish"
DEBU[0022] checking response                             status="200 OK"
DEBU[0022] sending request to Grafana Mimir API          method=GET url="http://localhost:8006/api/v1/upload/block/01GKN2TQV5A3D13852P6CKCS90/check"
DEBU[0022] checking response                             status="200 OK"
DEBU[0022] checked block upload state                    block=01GKN2TQV5A3D13852P6CKCS90 path=01GKN2TQV5A3D13852P6CKCS90 state=validating
DEBU[0042] sending request to Grafana Mimir API          method=GET url="http://localhost:8006/api/v1/upload/block/01GKN2TQV5A3D13852P6CKCS90/check"
DEBU[0042] checking response                             status="200 OK"
DEBU[0042] checked block upload state                    block=01GKN2TQV5A3D13852P6CKCS90 path=01GKN2TQV5A3D13852P6CKCS90 state=complete
INFO[0042] block uploaded successfully                   block=01GKN2TQV5A3D13852P6CKCS90 path=01GKN2TQV5A3D13852P6CKCS90
INFO[0042] finished uploading blocks                     already_exists=0 failed=0 succeeded=1

```

**Comment by aldernero** [comment]
Screenshot showing the concurrency gauge during the uploads
image

File: pkg/compactor/compactor.go [hunk 1]
```diff
@@ -365,6 +374,13 @@ func newMultitenantCompactor(
 		}),
 	}
 
+	promauto.With(registerer).NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "cortex_block_upload_current_validations",
```
Comment (pstibrany) [inline pkg/compactor/compactor.go:378]: nit: `cortex_block_upload_validations_in_progress` ?

File: pkg/compactor/block_upload.go [hunk 2]
```diff
@@ -291,6 +300,8 @@ func (c *MultitenantCompactor) validateAndCompleteBlockUpload(logger log.Logger,
 			c.periodicValidationUpdater(ctx, logger, blockID, userBkt, cancel, validationHeartbeatInterval)
 		}()
 
+		c.blockUploadValidations.Inc()
+		defer c.blockUploadValidations.Dec()
```
Comment (pstibrany) [inline pkg/compactor/block_upload.go:304]: I suggest to move this at the beginning of this method, just after log statement.

File: pkg/mimirtool/client/backfill.go [hunk 3]
```diff
@@ -100,11 +101,20 @@ func (c *MimirClient) backfillBlock(ctx context.Context, blockDir string, logctx
 		}
 	}
 
-	resp, err = c.doRequest(ctx, path.Join(endpointPrefix, url.PathEscape(blockID), finishBlockUpload), http.MethodPost, nil, -1)
-	if err != nil {
-		return errors.Wrap(err, "request to finish block upload failed")
+	for {
+		resp, err = c.doRequest(ctx, path.Join(endpointPrefix, url.PathEscape(blockID), finishBlockUpload), http.MethodPost, nil, -1)
+		if err != nil {
+			if strings.Contains(err.Error(), "429 Too Many Requests") {
```
Comment (pstibrany) [inline pkg/mimirtool/client/backfill.go:107]: Can we modify `checkResponse` (called from `c.doRequest`) to check for status code 429 and return an error that we can check for? We shouldn't be checking for status message string.

File: pkg/mimirtool/client/backfill.go [hunk 4]
```diff
@@ -100,11 +101,20 @@ func (c *MimirClient) backfillBlock(ctx context.Context, blockDir string, logctx
...
+				// wait and try again
+				logctx.WithField("error", err).Warning("will sleep and try again")
```
Comment (pstibrany) [inline pkg/mimirtool/client/backfill.go:109]: We can give user some context, eg. `failed to finish block upload and start block validation due to too many requests, will sleep and try again`. WDYT?

**Review by pstibrany — COMMENTED** [review]
Thank you! I like that you also added metric for validations in progress.

Can we also update [documentation for API endpoint](https://github.com/grafana/mimir/blob/236985ad6e3862aede7606fa04b0d81fda4d7aa1/docs/sources/mimir/references/http-api/index.md?plain=1#L1018-L1037) and mention that it can now return 429? This should also be mentioned in the changelog as `[CHANGE]` for the endpoint.

File: pkg/mimirtool/client/backfill.go (hunk 4 above)
Comment (pstibrany) [reply pkg/mimirtool/client/backfill.go:109]: I see that there is more context given in the log already (full response body), and my suggested message wouldn't fit into the format.

**Comment by pstibrany** [comment]
Very nice work on the PR. Thanks also for including example of output from mimirtool and screenshot from testing!

File: pkg/mimirtool/client/backfill.go [hunk 5]
```diff
@@ -100,11 +100,20 @@ func (c *MimirClient) backfillBlock(ctx context.Context, blockDir string, logctx
...
-	resp, err = c.doRequest(ctx, path.Join(endpointPrefix, url.PathEscape(blockID), finishBlockUpload), http.MethodPost, nil, -1)
-	if err != nil {
-		return errors.Wrap(err, "request to finish block upload failed")
+	for {
+		resp, err = c.doRequest(ctx, path.Join(endpointPrefix, url.PathEscape(blockID), finishBlockUpload), http.MethodPost, nil, -1)
+		if err != nil {
+			if errors.Is(err, errTooManyRequests) {
+				// wait and try again
+				logctx.WithField("error", err).Warning("will sleep and try again")
+				time.Sleep(sleepTime)
```
Comment (andyasp) [inline pkg/mimirtool/client/backfill.go:109]: dskit has a backoff mechanism [here](https://github.com/grafana/dskit/blob/727cbc22c1febfb5a77d024e407ebd01b207f74d/backoff/backoff.go#L26) that might be nice to use rather than directly sleeping.

File: pkg/compactor/block_upload.go [hunk 6]
```diff
@@ -104,6 +104,15 @@ func (c *MultitenantCompactor) FinishBlockUpload(w http.ResponseWriter, r *http.
 	}
 
 	if c.cfgProvider.CompactorBlockUploadValidationEnabled(tenantID) {
+		maxConcurrency := int64(c.compactorCfg.MaxBlockUploadValidationConcurrency)
+		if maxConcurrency > 0 && c.blockUploadValidations.Load() >= maxConcurrency {
```
Comment (andyasp) [inline pkg/compactor/block_upload.go:108]: Trusting `c.blockUploadValidations.Load()` allows a race where two can enter at once before either increments. Peeking with a `Load()` first is fine, but the actual check should `Inc()` and check if the return is <= maxConcurrency. If that check fails a decrement must follow though.

**Review by andyasp — COMMENTED** [review]
Nice work!

File: pkg/compactor/block_upload.go (hunk 6 above)
Comment (pstibrany) [reply pkg/compactor/block_upload.go:108]: Here's an example from ingester push method: https://github.com/grafana/mimir/blob/cfd882b23714deddaf71398e326ed9ff3d2611b5/pkg/ingester/ingester.go#L659-L668

**Comment by aldernero** [comment]
I re-ran some local tests, this time with `-compactor.max-block-upload-validation-concurrency=2` and started up 6 simultaneous uploads with test blocks:
```bash
❯ for block in $(ls -d 01G*);do echo $block;done
01GKGCXCHJPKV9J53E39RX8AS6
01GKJ1EG1T87PB8JFFVJ7RS07Y
01GKN1JNHT7PJCSVF8B1P772JS
01GKN2S8N18VZXQ7CCJEDHWDWD
01GKN2SS8BP9G2P7K2NWBKVC5W
01GKN2TQV5A3D13852P6CKCS90
❯ for block in $(ls -d 01G*);do (mimirtool backfill $block > upload-${block}.log 2>&1) &;done
[2] 117596
[3] 117597
[4] 117598
[5] 117599
[6] 117606
[7] 117610
```
I looked at `cortex_block_upload_validations_in_progess` to verify it was doing the right thing:
image

I verified from the MinIO container that all the blocks were there:
```bash
sh-4.4# ls /data/mimir-tsdb/anonymous/
01GKGCXCHJPKV9J53E39RX8AS6  01GKN2S8N18VZXQ7CCJEDHWDWD  01GWPYY6QMXPJFWX52BGX38EZJ  01GWPYYS30K33FHG1ASCB8H8EE
01GKJ1EG1T87PB8JFFVJ7RS07Y  01GKN2SS8BP9G2P7K2NWBKVC5W  01GWPYYC1PDY3H6GPWP6HDRBKB  bucket-index.json.gz
01GKN1JNHT7PJCSVF8B1P772JS  01GKN2TQV5A3D13852P6CKCS90  01GWPYYHHCWK3PJJVAE5EVMVWK  markers
```
Lastly I verified from the mimirtool output that all the clients finished successfully and retried when necessary:
```bash
❯ grep -e "succeeded" -e "too many" upload-*.log
upload-01GKGCXCHJPKV9J53E39RX8AS6.log:time="2023-03-29T09:13:46-06:00" level=warning msg="will sleep and try again" block=01GKGCXCHJPKV9J53E39RX8AS6 error="POST request to http://localhost:8006/api/v1/upload/block/01GKGCXCHJPKV9J53E39RX8AS6/finish failed: too many requests" path=01GKGCXCHJPKV9J53E39RX8AS6
upload-01GKGCXCHJPKV9J53E39RX8AS6.log:time="2023-03-29T09:14:12-06:00" level=info msg="finished uploading blocks" already_exists=0 failed=0 succeeded=1
upload-01GKJ1EG1T87PB8JFFVJ7RS07Y.log:time="2023-03-29T09:14:06-06:00" level=info msg="finished uploading blocks" already_exists=0 failed=0 succeeded=1
upload-01GKN1JNHT7PJCSVF8B1P772JS.log:time="2023-03-29T09:14:06-06:00" level=info msg="finished uploading blocks" already_exists=0 failed=0 succeeded=1
upload-01GKN2S8N18VZXQ7CCJEDHWDWD.log:time="2023-03-29T09:13:46-06:00" level=warning msg="will sleep and try again" block=01GKN2S8N18VZXQ7CCJEDHWDWD error="POST request to http://localhost:8006/api/v1/upload/block/01GKN2S8N18VZXQ7CCJEDHWDWD/finish failed: too many requests" path=01GKN2S8N18VZXQ7CCJEDHWDWD
upload-01GKN2S8N18VZXQ7CCJEDHWDWD.log:time="2023-03-29T09:14:11-06:00" level=info msg="finished uploading blocks" already_exists=0 failed=0 succeeded=1
upload-01GKN2SS8BP9G2P7K2NWBKVC5W.log:time="2023-03-29T09:13:46-06:00" level=warning msg="will sleep and try again" block=01GKN2SS8BP9G2P7K2NWBKVC5W error="POST request to http://localhost:8006/api/v1/upload/block/01GKN2SS8BP9G2P7K2NWBKVC5W/finish failed: too many requests" path=01GKN2SS8BP9G2P7K2NWBKVC5W
upload-01GKN2SS8BP9G2P7K2NWBKVC5W.log:time="2023-03-29T09:14:15-06:00" level=info msg="finished uploading blocks" already_exists=0 failed=0 succeeded=1
upload-01GKN2TQV5A3D13852P6CKCS90.log:time="2023-03-29T09:13:47-06:00" level=warning msg="will sleep and try again" block=01GKN2TQV5A3D13852P6CKCS90 error="POST request to http://localhost:8006/api/v1/upload/block/01GKN2TQV5A3D13852P6CKCS90/finish failed: too many requests" path=01GKN2TQV5A3D13852P6CKCS90
upload-01GKN2TQV5A3D13852P6CKCS90.log:time="2023-03-29T09:14:17-06:00" level=info msg="finished uploading blocks" already_exists=0 failed=0 succeeded=1
```

File: pkg/mimirtool/client/backfill.go [hunk 7]
```diff
@@ -100,11 +104,25 @@ func (c *MimirClient) backfillBlock(ctx context.Context, blockDir string, logctx
 		}
 	}
 
-	resp, err = c.doRequest(ctx, path.Join(endpointPrefix, url.PathEscape(blockID), finishBlockUpload), http.MethodPost, nil, -1)
-	if err != nil {
-		return errors.Wrap(err, "request to finish block upload failed")
+	retryConfig := backoff.Config{
+		MinBackoff: 5 * time.Second,
+		MaxBackoff: 1 * time.Minute,
+		MaxRetries: maxRetries,
```
[the remark this answers is not in the document]
Comment (andyasp) [reply pkg/mimirtool/client/backfill.go:110]: If the sleep time provided by the user should be respected then the backoff can be removed in favor of the direct sleep, apologies if I confused the issue by suggesting it. The benefit to using backoff would be respecting context cancellation and randomly distributing the checks to find when validation slots open since there is no correct choice for sleep time.

File: pkg/mimirtool/client/backfill.go [hunk 8]
```diff
@@ -15,13 +15,17 @@ import (
...
+	"github.com/grafana/dskit/backoff"
+
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 
 	"github.com/grafana/mimir/pkg/storage/tsdb/block"
 	"github.com/grafana/mimir/pkg/storage/tsdb/metadata"
 )
 
+const maxRetries = 10
```
[the remark this answers is not in the document]
Comment (aldernero) [reply pkg/mimirtool/client/backfill.go:27]: I'm going to go back to a regular sleep to be consistent with the CLI flag and what we do while waiting for validation to finish.

File: pkg/mimirtool/client/backfill.go [hunk 9]
```diff
@@ -100,11 +104,25 @@ func (c *MimirClient) backfillBlock(ctx context.Context, blockDir string, logctx
...
-	if err != nil {
-		return errors.Wrap(err, "request to finish block upload failed")
+	retryConfig := backoff.Config{
+		MinBackoff: 5 * time.Second,
+		MaxBackoff: 1 * time.Minute,
+		MaxRetries: maxRetries,
+	}
+	retry := backoff.New(ctx, retryConfig)
+
+	for retry.Ongoing() {
```
[the remark this answers is not in the document]
Comment (aldernero) [reply pkg/mimirtool/client/backfill.go:114]: I'm going to go back to a regular sleep to be consistent with the CLI flag and what we do while waiting for validation to finish.

File: pkg/mimirtool/client/backfill.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (aldernero) [reply pkg/mimirtool/client/backfill.go:110]: I'm going to go back to a regular sleep to be consistent with the CLI flag and what we do while waiting for validation to finish.

File: CHANGELOG.md [hunk 10]
```diff
@@ -64,6 +65,8 @@
 * [ENHANCEMENT] Compactor: validation of blocks uploaded via the TSDB block upload feature is now configurable on a per tenant basis: #4585
   * `-compactor.block-upload-validation-enabled` has been added, `compactor_block_upload_validation_enabled` can be used to override per tenant
   * `-compactor.block-upload.block-validation-enabled` was the previous global flag and has been removed
+* [ENHANCEMENT] OTLP: Add support for converting OTel exponential histograms to Prometheus native histograms. The ingestion of native histograms must be enabled, please set `-ingester.native-histograms-ingestion-enabled` to `true`. #4063
```
Comment (bboreham) [inline CHANGELOG.md:68]: This looks unrelated?

File: CHANGELOG.md [hunk 11]
```diff
@@ -14,6 +14,7 @@
 * [CHANGE] Ingester: the `cortex_ingester_tsdb_wal_replay_duration_seconds` metrics has been removed. #4465
 * [CHANGE] Query-frontend: use protobuf internal query result payload format by default. This feature is no longer considered experimental. #4557
 * [CHANGE] Ruler: reject creating federated rule groups while tenant federation is disabled. Previously the rule groups would be silently dropped during bucket sync. #4555
+* [CHANGE] Compactor: the `/api/v1/upload/block/{block}/finish` endpoint now returns a `429` status code when the compactor has reached the limit specified by `-compactor.max-block-upload-validation-concurrency`. #4598
```
Comment (bboreham) [inline CHANGELOG.md:17]: Nit: it wasn't immediately obvious why this relates to the compactor.  I realise that must be where the feature is implemented, but for the average person reading this something like "Backfill upload" might be better than "Compactor".

File: pkg/compactor/block_upload.go [hunk 12]
```diff
@@ -104,12 +104,30 @@ func (c *MultitenantCompactor) FinishBlockUpload(w http.ResponseWriter, r *http.
...
+		currentValidations := c.blockUploadValidations.Inc()
+		decreaseActiveValidationsInDefer := true
+		defer func() {
+			if decreaseActiveValidationsInDefer {
+				c.blockUploadValidations.Dec()
+			}
+		}()
+		if maxConcurrency > 0 && currentValidations > maxConcurrency {
+			err := httpError{
+				message:    "too many block upload validations in progress",
```
Comment (bboreham) [inline pkg/compactor/block_upload.go:117]: Should we include in the message what the limit is?

**Review by bboreham — COMMENTED** [review]
Code seems fine; I had a couple of nits, especially one line in CHANGELOG that looks unrelated.

File: CHANGELOG.md (hunk 11 above)
Comment (aldernero) [reply CHANGELOG.md:17]: Good point. I changed it to "TSDB Block Upload" since that's how it's described in our docs

**Review by 56quarters — APPROVED** [review]
